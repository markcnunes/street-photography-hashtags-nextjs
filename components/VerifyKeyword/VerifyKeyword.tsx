import styled from '@emotion/styled';
import Button from 'components/Button';
import KeywordsList from 'components/KeywordsList';
import { rgba } from 'polished';
import { useState } from 'react';
import { Category } from 'data/types';

type VerifyKeywordProps = Pick<Category, 'keywords'>;

/**
 * @desc outputs if keyword already exist
 */
const VerifyKeyword: React.FC<VerifyKeywordProps> = ({ keywords }) => {
  const [errorMessage, setErrorMessage] = useState<string | null>(null);
  const [keywordsAvalilable, setKeywordsAvalilable] = useState<
    VerifyKeywordProps['keywords']
  >([]);
  const [keywordsNotAvalilable, setKeywordsNotAvalilable] = useState<
    VerifyKeywordProps['keywords']
  >([]);
  const [inputValue, setInputValue] = useState<string>('');

  const handleAddOption = (e: React.FormEvent) => {
    e.preventDefault();

    let keyword = inputValue;

    // check if field is empty
    if (keyword.length < 1) {
      return setErrorMessage('Enter a valid value to add the item');
    }

    // The input accepts multiple keywords
    // It converts the string into array on every space
    const arrayKeywords: VerifyKeywordProps['keywords'] = keyword
      .split(/[ ]+/)
      .map((item: any) => item.trim());

    // should only start with #
    if (arrayKeywords.some((item: any) => !item.startsWith('#'))) {
      return setErrorMessage(`Keywords must start with #`);
    }
    // check for forbiden characters
    if (arrayKeywords.some((item: any) => !item.match(/^[a-zA-Z0-9#_ ]+$/i))) {
      return setErrorMessage(
        `Keywords can only start with # and be followoed by letters or _`,
      );
    }

    // create array of existing keywords
    const hasKeyword: VerifyKeywordProps['keywords'] = keywords.filter(
      (item) => {
        if (arrayKeywords.indexOf(item) > -1) {
          return item;
        }
        return null;
      },
    );

    // filter new keywords from hasKeyword
    const newKeywords: VerifyKeywordProps['keywords'] = arrayKeywords
      .filter((x: string) => !hasKeyword.includes(x))
      .concat(hasKeyword.filter((x) => !arrayKeywords.includes(x)));

    // output new keywords
    if (newKeywords.length > 0) {
      setKeywordsAvalilable(newKeywords);
    }

    // output existing keywords
    if (hasKeyword.length > 0) {
      setKeywordsNotAvalilable(hasKeyword);
    }

    // reset state to defaul values
    setInputValue('');
    if (errorMessage !== typeof undefined) {
      setErrorMessage(null);
    }
  };

  const handleTextChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    setInputValue(e.target.value);
  };

  return (
    <StyledVerifyKeyword>
      {errorMessage && <p>{errorMessage}</p>}
      <form onSubmit={handleAddOption}>
        <input
          className='add-option__input'
          type='text'
          name='option'
          value={inputValue}
          onChange={handleTextChange}
          placeholder='Type a keyword. Use space for multiple keywords'
        />

        <Button type='submit' variant='outlined' color='white'>
          Verify keyword
        </Button>
      </form>
      {keywordsNotAvalilable.length > 0 && (
        <div className='keywords-list'>
          <h3>
            {keywordsNotAvalilable.length > 1
              ? 'These keywords '
              : 'This keyword '}
            already exist{keywordsNotAvalilable.length > 1 ? 's' : ''}:
          </h3>
          <KeywordsList keywords={keywordsNotAvalilable} />
        </div>
      )}

      {keywordsAvalilable.length > 0 && (
        <div className='keywords-list'>
          <h3>New Keyword{keywordsAvalilable.length > 1 && 's'}:</h3>
          <KeywordsList keywords={keywordsAvalilable} />
        </div>
      )}
    </StyledVerifyKeyword>
  );
};
export default VerifyKeyword;

const StyledVerifyKeyword = styled('div')(({ theme }) => ({
  display: 'inline-flex',
  flexWrap: 'wrap',
  alignItems: 'center',
  marginTop: 20,
  // From
  form: {
    display: 'inline-flex',
    width: '100%',
    input: {
      border: 'none',
      color: theme.colors.black,
      textAlign: 'center',
      flex: '1 0 0',
      padding: '10px 20px',
      borderRadius: '4px',
    },
    button: {
      marginLeft: 10,
    },
  },
  // Keywords list
  '.keywords-list': {
    width: '100%',
    marginTop: 20,
    h3: {
      color: theme.colors.white,
    },
    '.sidebar': {
      flexDirection: 'row',
      alignItems: 'center',
      right: 0,
      top: 0,
      padding: '0 20px',
      height: '100%',
      background: rgba(theme.colors.muted, 0.1),
      borderLeft: `1px solid ${rgba(theme.colors.muted, 0.2)}`,
      borderRadius: '0 5px 5px 0',
      '& > *': {
        margin: 5,
      },
      '.counter+.copy': {
        marginTop: 5,
      },
    },
  },
}));
