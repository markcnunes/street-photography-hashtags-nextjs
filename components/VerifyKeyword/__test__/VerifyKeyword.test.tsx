import React from 'react';
import { render, screen } from 'test/setupTests';
import userEvent from '@testing-library/user-event';
import VerifyKeyword from '../VerifyKeyword';
import mockData from '../__mock__/mockData';

const { keywords } = mockData;

describe('<VerifyKeyword />', () => {
  it('matchs snapshot', () => {
    const { asFragment } = render(<VerifyKeyword keywords={keywords} />);
    expect(asFragment()).toMatchSnapshot();
  });
  it('display new keyword', () => {
    render(<VerifyKeyword keywords={keywords} />);
    const input = screen.getByPlaceholderText(
      /Type a keyword. Use space for multiple keywords/i,
    );

    // add some keyword
    userEvent.type(input, '#newKeyword');
    expect(input).toHaveValue('#newKeyword');

    // click to open modal
    userEvent.click(screen.getByText('Verify keyword'));

    expect(screen.getByText(/New Keyword/i)).toBeInTheDocument();
    expect(screen.getByText('#newKeyword')).toBeInTheDocument();
  });
  it('displays if keyword exist', () => {
    render(<VerifyKeyword keywords={keywords} />);
    const input = screen.getByPlaceholderText(
      /Type a keyword. Use space for multiple keywords/i,
    );

    // add some keyword
    userEvent.type(input, '#test1');
    expect(input).toHaveValue('#test1');

    // click to open modal
    userEvent.click(screen.getByText('Verify keyword'));

    expect(screen.getByText(/This keyword already exist/i)).toBeInTheDocument();
    expect(screen.getByText('#test1')).toBeInTheDocument();
  });
  describe('Check rules', () => {
    it(`only accepts keywords starting with #`, () => {
      render(<VerifyKeyword keywords={keywords} />);
      const input = screen.getByPlaceholderText(
        /Type a keyword. Use space for multiple keywords/i,
      );

      // add some keyword
      userEvent.type(input, '@wrog');
      expect(input).toHaveValue('@wrog');

      // click to open modal
      userEvent.click(screen.getByText('Verify keyword'));

      expect(
        screen.getByText(/Keywords must start with #/i),
      ).toBeInTheDocument();
    });
    it(`renders keywords with _`, () => {
      render(<VerifyKeyword keywords={keywords} />);
      const input = screen.getByPlaceholderText(
        /Type a keyword. Use space for multiple keywords/i,
      );

      // add some keyword
      userEvent.type(input, '#some_keyword');
      expect(input).toHaveValue('#some_keyword');

      // click to open modal
      userEvent.click(screen.getByText('Verify keyword'));

      expect(screen.getByText(/New Keyword/i)).toBeInTheDocument();
      expect(screen.getByText('#some_keyword')).toBeInTheDocument();
    });
    it(`doesn't accept special character`, () => {
      render(<VerifyKeyword keywords={keywords} />);
      const input = screen.getByPlaceholderText(
        /Type a keyword. Use space for multiple keywords/i,
      );

      // add some keyword
      userEvent.type(input, '#notAllows@');
      expect(input).toHaveValue('#notAllows@');

      // click to open modal
      userEvent.click(screen.getByText('Verify keyword'));

      expect(
        screen.getByText(
          /Keywords can only start with # and be followoed by letters or _/i,
        ),
      ).toBeInTheDocument();
    });
  });
});
