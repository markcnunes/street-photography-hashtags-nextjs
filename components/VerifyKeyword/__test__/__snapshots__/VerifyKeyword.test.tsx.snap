// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<VerifyKeyword /> matchs snapshot 1`] = `
<DocumentFragment>
  .emotion-0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-flex-wrap: wrap;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin-top: 20px;
}

.emotion-0 form {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  width: 100%;
}

.emotion-0 form input {
  border: none;
  color: #1c1c27;
  text-align: center;
  -webkit-flex: 1 0 0;
  -ms-flex: 1 0 0;
  flex: 1 0 0;
  padding: 10px 20px;
  border-radius: 4px;
}

.emotion-0 form button {
  margin-left: 10px;
}

.emotion-0 .keywords-list {
  width: 100%;
  margin-top: 20px;
}

.emotion-0 .keywords-list h3 {
  color: #fff;
}

.emotion-0 .keywords-list .sidebar {
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  right: 0;
  top: 0;
  padding: 0 20px;
  height: 100%;
  background: rgba(83,96,117,0.1);
  border-left: 1px solid rgba(83,96,117,0.2);
  border-radius: 0 5px 5px 0;
}

.emotion-0 .keywords-list .sidebar>* {
  margin: 5px;
}

.emotion-0 .keywords-list .sidebar .counter+.copy {
  margin-top: 5px;
}

.emotion-2 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: auto;
  -ms-flex-pack: auto;
  -webkit-justify-content: auto;
  justify-content: auto;
  border: 0;
  outline: 0;
  cursor: pointer;
  width: auto;
  opacity: 1;
  border-radius: 5px;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-transition: background 0.3s ease;
  transition: background 0.3s ease;
  color: #1c1c27;
  background: #fff;
  font-size: 16px;
  padding: 8px 16px;
  border-radius: 4px;
  border: 1px solid currentColor;
  background: transparent;
  color: #fff;
}

.emotion-2:disabled {
  cursor: auto;
}

.emotion-2:hover {
  background: #fff;
}

.emotion-2:disabled {
  color: rgba(83,96,117,0.5);
}

.emotion-2:hover {
  background: rgba(255,255,255,0.05);
}

.emotion-2:hover:disabled {
  background: transparent;
  color: rgba(83,96,117,0.5);
}

<div
    class="emotion-0 emotion-1"
  >
    <form>
      <input
        class="add-option__input"
        name="option"
        placeholder="Type a keyword. Use space for multiple keywords"
        type="text"
        value=""
      />
      <button
        class="emotion-2 emotion-3"
        color="white"
        type="submit"
      >
        Verify keyword
      </button>
    </form>
  </div>
</DocumentFragment>
`;
